src/core/Carousel/Carousel.tsx

```diff
@@ -0,0 +1,494 @@
+  }
+
+  const { currentIndex, setSlideCount, idPrefix } = context;
+  const sliderRef = React.useRef<HTMLOListElement>(null);
+  const refs = useMergedRefs(sliderRef, ref);
+
+  React.useLayoutEffect(() => {
+    setSlideCount(React.Children.count(props.children));
+  }, [props.children, setSlideCount]);
+
+  const items = React.useMemo(
+    () =>
+      React.Children.map(props.children, (child, index) =>
+        React.isValidElement(child)
+          ? React.cloneElement(child, {
+              id: `${idPrefix}--slide-${index}`,
+              index,
+            })
+          : child,
+      ) ?? [],
+    [props.children, idPrefix],
+  ) as React.ReactNode[];
+
+  React.useEffect(() => {
+    if (!sliderRef.current) {
+      return;
+    }
+
+    sliderRef.current.children.item(currentIndex)?.scrollIntoView();
+  }, [currentIndex]);
+
+  return (
+    <ol
+      aria-live='polite'
+      {...props}
+      ref={refs}
+      className={cx('iui-carousel-slider', props.className)}
+    >
+      {items}
+    </ol>
+  );
+});
+
+/**
+ * CarouselSlide is used for the actual slide content. The content can be specified through `children`.
+ *
+ * It is recommended that the slide content bring its own dimensions (esp. height) and that
+ * the dimensions should be the same for all slides.
+ */
+const CarouselSlide = React.forwardRef<
+  HTMLLIElement,
+  { index?: number } & React.ComponentPropsWithoutRef<'li'>
+>(({ index, ...props }, ref) => {
+  const { className, children, ...rest } = props;
+
+  const context = React.useContext(CarouselContext);
+  if (!context) {
+    throw new Error('CarouselSlider must be used within Carousel');
+  }
+
+  const { currentIndex, isManuallyUpdating, setCurrentIndex } = context;
+
+  const updateActiveIndex = React.useCallback(() => {
+    if (!isManuallyUpdating.current) {
+      setCurrentIndex((prev) => (prev !== index ? index : prev));
+    }
+    if (currentIndex === index) {
+      isManuallyUpdating.current = false;
+    }
+  }, [currentIndex, index, isManuallyUpdating, setCurrentIndex]);
+
+  const intersectionRef = useIntersection(
+    updateActiveIndex,
+    { threshold: 0.5 },
+    false,
+  );
+
+  const refs = useMergedRefs(intersectionRef, ref);
+
+  return (
+    <li
+      className={cx('iui-carousel-slider-item', className)}
+      role='tabpanel'
+      aria-roledescription='slide'
+      ref={refs}
+      {...rest}
+    >
+      {children}
+    </li>
+  );
+});
+
+/**
+ * The CarouselNavigation component by default consists of the previous/next slide buttons
+ * shown on the left and right, and the CarouselDots component shown in the middle.
+ *
+ * `children` can be specified to override what is shown in this section.
+ */
+const CarouselNavigation = React.forwardRef<
+  HTMLElement,
+  React.ComponentPropsWithoutRef<'nav'>
+>((props, ref) => {
+  const { className, children, ...rest } = props;
+
+  const context = React.useContext(CarouselContext);
```

Comment:
Seems like this could be moved to a hook.

Reply:
Yeah, but then you lose the more specific error messages where you can use the name of the component

Reply:
Have component name as a parameter then? Or user could just look at call stack and see that Error was called from hook that was called from a component.